Invalid query leads to exception

I entered:

mcs search :quarkus-parent

and got:
```

Searching for :quarkus-parent...
Error parsing search response: Unrecognized token 'Solr': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream); line: 1, column: 6]
java.io.UncheckedIOException: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Solr': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream); line: 1, column: 6]
        at it.mulders.mcs.common.SearchResponseBodyHandler.lambda$toSupplierOfSearchResponse$0(SearchResponseBodyHandler.java:36)
        at it.mulders.mcs.search.SearchClient.performSearch(SearchClient.java:77)
        at it.mulders.mcs.search.SearchClient.singularSearch(SearchClient.java:49)
        at it.mulders.mcs.search.SearchCommandHandler.performCoordinateSearch(SearchCommandHandler.java:52)
        at it.mulders.mcs.search.SearchCommandHandler.search(SearchCommandHandler.java:21)
        at it.mulders.mcs.cli.Cli$SearchCommand.call(Cli.java:49)
        at it.mulders.mcs.cli.Cli$SearchCommand.call(Cli.java:32)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at it.mulders.mcs.App.doMain(App.java:17)
        at it.mulders.mcs.App.main(App.java:10)
Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Solr': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream); line: 1, column: 6]
        at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2391)
        at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:745)
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._reportInvalidToken(UTF8StreamJsonParser.java:3634)
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._handleUnexpectedValue(UTF8StreamJsonParser.java:2734)
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._nextTokenNotInObject(UTF8StreamJsonParser.java:902)
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:794)
        at com.fasterxml.jackson.jr.ob.JSON._initForReading(JSON.java)
        at com.fasterxml.jackson.jr.ob.JSON.mapFrom(JSON.java)
        at it.mulders.mcs.common.SearchResponseBodyHandler.lambda$toSupplierOfSearchResponse$0(SearchResponseBodyHandler.java:32)
        ... 16 more
```

--- a/pom.xml
+++ b/pom.xml
@@ -267,6 +267,10 @@
         </dependency>
         <dependency>
             <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
             <artifactId>junit-jupiter-engine</artifactId>
             <scope>test</scope>
         </dependency>

--- a/src/main/java/it/mulders/mcs/search/CoordinateQuery.java
+++ b/src/main/java/it/mulders/mcs/search/CoordinateQuery.java
@@ -2,6 +2,8 @@ package it.mulders.mcs.search;
 
 import java.net.URLEncoder;
 import java.nio.charset.StandardCharsets;
+import java.util.LinkedList;
+import java.util.List;
 
 import static it.mulders.mcs.search.Constants.DEFAULT_MAX_SEARCH_RESULTS;
 
@@ -13,12 +15,11 @@ public record CoordinateQuery (
 ) implements SearchQuery {
     @Override
     public String toSolrQuery() {
-        String query;
-        if (version == null) {
-            query = String.format("g:%s AND a:%s", groupId, artifactId);
-        } else {
-            query = String.format("g:%s AND a:%s AND v:%s", groupId, artifactId, version);
-        }
+        final List<String> parts = new LinkedList<>();
+        if (!groupId.isBlank()) parts.add("g:%s".formatted(groupId));
+        if (!artifactId.isBlank()) parts.add("a:%s".formatted(artifactId));
+        if (!version.isBlank()) parts.add("v:%s".formatted(version));
+        var query = String.join(" AND ", parts);
 
         return String.format("q=%s&core=gav&start=%d&rows=%d",
                 URLEncoder.encode(query, StandardCharsets.UTF_8), 0, searchLimit());
@@ -35,9 +36,14 @@ public record CoordinateQuery (
         }
 
         public Builder(String groupId, String artifactId, String version) {
-            this.groupId = groupId;
-            this.artifactId = artifactId;
-            this.version = version;
+            this.groupId = sanitise(groupId);
+            this.artifactId = sanitise(artifactId);
+            this.version = sanitise(version);
+        }
+
+        private String sanitise(String input) {
+            if (input == null) return "";
+            return input;
         }
 
         @Override

--- a/src/main/java/it/mulders/mcs/search/SearchQuery.java
+++ b/src/main/java/it/mulders/mcs/search/SearchQuery.java
@@ -10,6 +10,7 @@ public sealed interface SearchQuery permits CoordinateQuery, ClassnameQuery, Wil
         if (isCoordinateSearch) {
             var parts = query.split(":");
             switch (parts.length) {
+                case 1: return new CoordinateQuery.Builder(parts[0], null);
                 case 2: return new CoordinateQuery.Builder(parts[0], parts[1]);
                 case 3: return new CoordinateQuery.Builder(parts[0], parts[1], parts[2]);
                 default:

--- a/src/test/java/it/mulders/mcs/search/CoordinateQueryTest.java
+++ b/src/test/java/it/mulders/mcs/search/CoordinateQueryTest.java
@@ -15,7 +15,7 @@ class CoordinateQueryTest implements WithAssertions {
     class ToSolrQueryTest {
         @Test
         void solr_query_should_contain_groupId() {
-            var query = new CoordinateQuery("foo", "bar", null, Constants.DEFAULT_MAX_SEARCH_RESULTS);
+            var query = new CoordinateQuery("foo", "bar", "", Constants.DEFAULT_MAX_SEARCH_RESULTS);
 
             var solrQuery = query.toSolrQuery();
 
@@ -24,7 +24,7 @@ class CoordinateQueryTest implements WithAssertions {
 
         @Test
         void solr_query_should_contain_artifactId() {
-            var query = new CoordinateQuery("foo", "bar", null, Constants.DEFAULT_MAX_SEARCH_RESULTS);
+            var query = new CoordinateQuery("foo", "bar", "", Constants.DEFAULT_MAX_SEARCH_RESULTS);
 
             var solrQuery = query.toSolrQuery();
 
@@ -42,7 +42,7 @@ class CoordinateQueryTest implements WithAssertions {
 
         @Test
         void solr_query_should_contain_start() {
-            var query = new CoordinateQuery("foo", "bar", null, Constants.DEFAULT_MAX_SEARCH_RESULTS);
+            var query = new CoordinateQuery("foo", "bar", "", Constants.DEFAULT_MAX_SEARCH_RESULTS);
 
             var solrQuery = query.toSolrQuery();
 

--- a/src/test/java/it/mulders/mcs/search/SearchCommandHandlerTest.java
+++ b/src/test/java/it/mulders/mcs/search/SearchCommandHandlerTest.java
@@ -36,7 +36,7 @@ class SearchCommandHandlerTest implements WithAssertions {
         public Result<SearchResponse> search(final SearchQuery query) {
             if (query instanceof WildcardSearchQuery) {
                 return new Result.Success<>(new SearchResponse(null, wildcardResponse));
-            } else if (query instanceof CoordinateQuery cq && cq.version() == null) {
+            } else if (query instanceof CoordinateQuery cq && cq.version().isBlank()) {
                 return new Result.Success<>(new SearchResponse(null, twoPartCoordinateResponse));
             } else {
                 return new Result.Success<>(new SearchResponse(null, threePartCoordinateResponse));

--- a/src/test/java/it/mulders/mcs/search/SearchQueryTest.java
+++ b/src/test/java/it/mulders/mcs/search/SearchQueryTest.java
@@ -1,0 +1,69 @@
+package it.mulders.mcs.search;
+
+import org.assertj.core.api.WithAssertions;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+
+class SearchQueryTest implements WithAssertions {
+    @Nested
+    class CoordinateQueryTest {
+        @Test
+        void should_build_query_with_groupId_and_artifactId() {
+            var query = SearchQuery.search("org.codehaus.plexus:plexus-utils").build();
+            assertThat(query).isInstanceOf(CoordinateQuery.class).satisfies(q -> {
+                assertThat(((CoordinateQuery) q).groupId()).isEqualTo("org.codehaus.plexus");
+                assertThat(((CoordinateQuery) q).artifactId()).isEqualTo("plexus-utils");
+                assertThat(((CoordinateQuery) q).version()).isNullOrEmpty();
+            });
+        }
+
+        @Test
+        void should_build_query_with_groupId_and_artifactId_and_version() {
+            var query = SearchQuery.search("org.codehaus.plexus:plexus-utils:3.4.1").build();
+            assertThat(query).isInstanceOf(CoordinateQuery.class).satisfies(q -> {
+                assertThat(((CoordinateQuery) q).groupId()).isEqualTo("org.codehaus.plexus");
+                assertThat(((CoordinateQuery) q).artifactId()).isEqualTo("plexus-utils");
+                assertThat(((CoordinateQuery) q).version()).isEqualTo("3.4.1");
+            });
+        }
+
+        @Test
+        void should_build_query_with_only_groupId() {
+            var query = SearchQuery.search("org.codehaus.plexus:").build();
+            assertThat(query).isInstanceOf(CoordinateQuery.class).satisfies(q -> {
+                assertThat(((CoordinateQuery) q).groupId()).isEqualTo("org.codehaus.plexus");
+                assertThat(((CoordinateQuery) q).artifactId()).isNullOrEmpty();
+                assertThat(((CoordinateQuery) q).version()).isNullOrEmpty();
+            });
+        }
+
+        @Test
+        void should_build_query_with_only_artifactId() {
+            var query = SearchQuery.search(":plexus-utils").build();
+            assertThat(query).isInstanceOf(CoordinateQuery.class).satisfies(q -> {
+                assertThat(((CoordinateQuery) q).groupId()).isNullOrEmpty();
+                assertThat(((CoordinateQuery) q).artifactId()).isEqualTo("plexus-utils");
+                assertThat(((CoordinateQuery) q).version()).isNullOrEmpty();
+            });
+        }
+
+        @ParameterizedTest
+        @CsvSource(textBlock = """
+                org.codehaus.plexus:plexus-utils,g:org.codehaus.plexus AND a:plexus-utils
+                org.codehaus.plexus:plexus-utils:3.4.1,g:org.codehaus.plexus AND a:plexus-utils AND v:3.4.1
+                org.codehaus.plexus:,g:org.codehaus.plexus
+                :plexus-utils,a:plexus-utils
+                """)
+        void should_construct_valid_solr_query(String input, String solrQuery) {
+            var result = SearchQuery.search(input).build()
+                    .toSolrQuery();
+
+            assertThat(result).contains("q=" + URLEncoder.encode(solrQuery, StandardCharsets.UTF_8));
+        }
+    }
+}

